### PR TITLE
Add kubernetes module for aws

### PR DESCRIPTION
**Pull request by tei-k**
# Description
https://scalar-labs.atlassian.net/browse/DLT-7400

# Done
- Add Kubernetes module for AWS
- Only support node group

# Confirm

- Terraform apply
```
$ cd example/aws/kubernetes
$ terraform apply -var-file=example.tfvars
```
- Copy configuration files (`scalar-k8/conf`)
  - inventory.ini
  - kube_config (Copy from `${eks_cluster_name}_kubeconfig`)
- Fix `scalar-k8/conf/scalardl-custom-values.yaml`
```
-      service.beta.kubernetes.io/azure-load-balancer-internal: "true"
-      service.beta.kubernetes.io/azure-load-balancer-internal-subnet: "k8s_ingress"
+      service.beta.kubernetes.io/aws-load-balancer-type: clb
+      service.beta.kubernetes.io/aws-load-balancer-internal: "true"
+      # service.beta.kubernetes.io/azure-load-balancer-internal: "true"
+      # service.beta.kubernetes.io/azure-load-balancer-internal-subnet: "k8s_ingress"
```
- ansible-playbook (`scalar-k8/`)
```
$ ansible-playbook -i ${SCALAR_K8S_CONFIG_DIR}/inventory.ini playbooks/playbook-install-tools.yml
$ ansible-playbook -i ${SCALAR_K8S_CONFIG_DIR}/inventory.ini playbooks/playbook-deploy-scalardl.yml
$ ansible-playbook -i ${SCALAR_K8S_CONFIG_DIR}/inventory.ini playbooks/playbook-deploy-prometheus.yml
```
- Get resources
```
$ k get pod,svc -o wide
```

# Ref
- https://github.com/terraform-aws-modules/terraform-aws-eks
- https://github.com/cloudposse/terraform-aws-eks-cluster

### CODE REVIEW

File: modules/aws/kubernetes/iam_cluster.tf [hunk 1]
```diff
@@ -0,0 +1,96 @@
+resource "aws_iam_role" "eks_cluster" {
```
Comment (feeblefakie) [inline modules/aws/kubernetes/iam_cluster.tf:1]: Do we need to add a subsection for privileges in EKS deployment?
https://github.com/scalar-labs/scalar-terraform/blob/master/docs/CloudPrivileges.md

**Review by feeblefakie — COMMENTED** [review]
Overall looking good. Left some suggestions and comments.

File: modules/aws/kubernetes/vars.tf [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+variable "kubernetes_cluster" {
+  type        = map
+  default     = {}
+  description = "Custom definition kubernetes properties that include name of the cluster, kubernetes version, etc.."
+}
+
+variable "kubernetes_node_groups" {
+  description = "Map of map of node groups to create"
```
[the remark this answers is not in the document]
Comment (tei-k) [reply modules/aws/kubernetes/vars.tf:14]: Yes, It's a map of maps.

File: modules/aws/kubernetes/iam_cluster.tf (hunk 1 above)
Comment (tei-k) [reply modules/aws/kubernetes/iam_cluster.tf:1]: Oh, We need some privileges like below.
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/iam-permissions.md

File: examples/aws/kubernetes/example.tfvars [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Optional
```
Comment (ymorimo) [inline examples/aws/kubernetes/example.tfvars:1]: As with the other examples, the default region should be here at the top of `example.tfvars`.
 (I didn't use the default `ap-northeast-1` region and I got lost when `terraform apply` couldn't find the VPC that the network module created.)

```suggestion
region = "ap-northeast-1"

# Optional
```

Comment (tei-k) [reply examples/aws/kubernetes/example.tfvars]: Good point! 👍 
Also added `backend.tf.s3` and `remote.tf.s3` in 41f7f45

**Comment by ymorimo** [comment]
@tei-k Question.

I got the following kube_config output after `terraform apply`,

```yaml
clusters:
- cluster:
    server: https://D0BE5C2FAD77ED66ED29817523BB7182.yl4.us-west-1.eks.amazonaws.com
    certificate-authority-data: LS0tLS1CRUdJTiBDRV...
```

but the server host in that file resolves to private IP addresses. Because of this `kubectl` fails to communicate with the endpoint.

```console
$ host d0be5c2fad77ed66ed29817523bb7182.yl4.us-west-1.eks.amazonaws.com
d0be5c2fad77ed66ed29817523bb7182.yl4.us-west-1.eks.amazonaws.com has address 10.42.1.144
d0be5c2fad77ed66ed29817523bb7182.yl4.us-west-1.eks.amazonaws.com has address 10.42.41.117
```

Can you tell what is happening?

**Comment by tei-k** [comment]
@ymorimo 

>but the server host in that file resolves to private IP addresses. Because of this kubectl fails to communicate with the endpoint.

Good point! 👍 Sorry for missing of explanation on this point.

Because `cluster_endpoint_public_access` is false in the current default setting, So cannot access this endpoint from local. 
To operate from the bastion we need to set up authentication manually.

Can you first create environment with `cluster_endpoint_public_access = "true"` ?

example.tfvars
``` 
# cluster_endpoint_private_access      = "true"
cluster_endpoint_public_access       = "true"
```

**Review by feeblefakie — APPROVED** [review]
LGTM!

Please create a PR for the update in `scalardl-custom-values.yaml` once this PR is merged.
Also, please write a getting started doc once Yusuke's doc reorganization is completed.

**Review by ymorimo — APPROVED** [review]
It worked. Thanks. LGTM!

**Pull request merged**